File: app/src/main/java/com/kota/Bahamut/DataModels/ArticleTempStore.java
package com.kota.Bahamut.DataModels;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.util.Iterator;
import java.util.Vector;

public class ArticleTempStore {
    public static final int version = 1;
    private final Context _context;
    private String _file_path = null;
    public Vector<ArticleTemp> articles = new Vector<>();

    public static void upgrade(Context context, String aFilePath) {
        new ArticleTempStore(context, aFilePath).load();
    }

    public ArticleTempStore(Context context, String aFilePath) {
        this._context = context;
        this._file_path = aFilePath;
        for (int i=0;i<10;i++) {
            this.articles.add(new ArticleTemp());
        }
    }

    public ArticleTempStore(Context context) {
        this._context = context;
        for (int i=0;i<10;i++) {
            this.articles.add(new ArticleTemp());
        }
        load();
    }

    public void load() {
        System.out.println("load article store from file");
        boolean load_from_file = false;
        if (this._file_path != null && this._file_path.length() > 0) {
            File file = new File(this._file_path);
            if (file.exists()) {
                try {
                    InputStream file_input_stream = new FileInputStream(file);
                    ObjectInputStream input_stream = new ObjectInputStream(file_input_stream);
                    importFromStream(input_stream);
                    input_stream.close();
                    file_input_stream.close();
                    file.delete();
                    load_from_file = true;
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        if (load_from_file) {
            store();
        }
        if (!load_from_file && this._context != null) {
            try {
                String save_data = this._context.getSharedPreferences("article_temp", 0).getString("save_data", "");
                if (save_data.length() > 0) {
                    importFromJSON(new JSONObject(save_data));
                }
            } catch (Exception e2) {
                e2.printStackTrace();
            }
        }
    }

    public void store() {
        if (this._context != null) {
            try {
                SharedPreferences perf = this._context.getSharedPreferences("article_temp", 0);
                perf.edit().putString("save_data", exportToJSON().toString()).commit();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void importFromJSON(JSONObject obj) throws JSONException {
        JSONArray data = obj.getJSONArray("data");
        this.articles.clear();
        for (int i = 0; i < data.length(); i++) {
            JSONObject item_data = data.getJSONObject(i);
            ArticleTemp temp = new ArticleTemp();
            temp.importFromJSON(item_data);
            this.articles.add(temp);
        }
    }

    public JSONObject exportToJSON() throws JSONException {
        JSONObject obj = new JSONObject();
        JSONArray save_data = new JSONArray();
        for (ArticleTemp article : this.articles) {
            save_data.put(article.exportToJSON());
        }
        obj.put("data", save_data);
        return obj;
    }

    public void importFromStream(ObjectInputStream aStream) throws IOException {
        aStream.readInt();
        int size = aStream.readInt();
        for (int i = 0; i < size; i++) {
            ArticleTemp article = new ArticleTemp();
            article.importFromStream(aStream);
            this.articles.add(article);
        }
    }
}

File: app/src/main/java/com/kota/Bahamut/dataModels/ArticleTempStore.kt
package com.kota.Bahamut.dataModels

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.io.ObjectInputStream
import java.util.Vector
import androidx.core.content.edit

/*
    戰巴哈專用暫存檔
    0~4 文章使用
    8  mail
    9  發文
 */
class ArticleTempStore {
    private val context: Context?
    private var filePath: String? = null
    @JvmField
    var articles: Vector<ArticleTemp> = Vector<ArticleTemp>()

    constructor(context: Context?, aFilePath: String?) {
        this.context = context
        this.filePath = aFilePath
        for (i in 0..9) {
            this.articles.add(ArticleTemp())
        }
    }

    constructor(context: Context?) {
        this.context = context
        for (i in 0..9) {
            this.articles.add(ArticleTemp())
        }
        load()
    }

    fun load() {
        println("load article store from file")
        var loadFromFile = false
        if (this.filePath != null && this.filePath?.isNotEmpty()) {
            val file = File(this.filePath!!)
            if (file.exists()) {
                try {
                    val fileInputStream: InputStream = FileInputStream(file)
                    val inputStream = ObjectInputStream(fileInputStream)
                    importFromStream(inputStream)
                    inputStream.close()
                    fileInputStream.close()
                    file.delete()
                    loadFromFile = true
                } catch (e: IOException) {
                    Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
                }
            }
        }
        if (loadFromFile) {
            store()
        }
        if (!loadFromFile && this.context != null) {
            try {
                val saveData: String = this.context.getSharedPreferences("article_temp", 0)
                    .getString("save_data", "")!!
                if (saveData.isNotEmpty()) {
                    importFromJSON(JSONObject(saveData))
                }
            } catch (e2: Exception) {
                e2.printStackTrace()
            }
        }
    }

    fun store() {
        if (this.context != null) {
            try {
                val perf = this.context.getSharedPreferences("article_temp", 0)
                perf.edit(commit = true) { putString("save_data", exportToJSON().toString()) }
            } catch (e: Exception) {
                Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
            }
        }
    }

    @Throws(JSONException::class)
    fun importFromJSON(obj: JSONObject) {
        val data = obj.getJSONArray("data")
        this.articles.clear()
        for (i in 0..<data.length()) {
            val itemData = data.getJSONObject(i)
            val temp = ArticleTemp()
            temp.importFromJSON(itemData)
            this.articles.add(temp)
        }
    }

    @Throws(JSONException::class)
    fun exportToJSON(): JSONObject {
        val obj = JSONObject()
        val saveData = JSONArray()
        for (article in this.articles) {
            saveData.put(article.exportToJSON())
        }
        obj.put("data", saveData)
        return obj
    }

    @Throws(IOException::class)
    fun importFromStream(aStream: ObjectInputStream) {
        aStream.readInt()
        val size = aStream.readInt()
        for (i in 0..<size) {
            val article = ArticleTemp()
            article.importFromStream(aStream)
            this.articles.add(article)
        }
    }

    companion object {
        const val VERSION: Int = 1
        @JvmStatic
        fun upgrade(context: Context?, aFilePath: String?) {
            ArticleTempStore(context, aFilePath).load()
        }
    }
}
